src/stack.h

```diff
@@ -0,0 +1,81 @@
+
+#include <stdio.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef DEBUG
+    #define LOG(args...) fprintf(stderr, args);
+#else
+    #define LOG(args...)
+#endif
+
+typedef struct ScannerStack {
+  unsigned int stack[256];
+  int top;
+} ScannerStack;
+
+ScannerStack* createStack() {
+  ScannerStack* ptr = (ScannerStack*) malloc(sizeof(ScannerStack));
+
+  ptr -> top = 0;
+  memset(ptr -> stack, 256, (0));
+
+  return ptr;
+}
+
+bool isEmptyStack(ScannerStack *stack) { return stack->top == 0; }
+
+int peekStack(ScannerStack *stack) {
+  return isEmptyStack(stack) ? -1 : stack->stack[stack->top - 1];
+}
+
+void pushStack(ScannerStack *stack, unsigned int value) {
+  stack->top++;
+  stack->stack[stack->top - 1] = value;
+}
+
+int popStack(ScannerStack *stack) {
+  if (isEmptyStack(stack))
+    return -1;
+  else {
+    int result = peekStack(stack);
+    stack->top--;
+
+    return result;
+  }
+}
+
+void printStack(ScannerStack *stack, char *msg) {
+  LOG("%s Stack[top = %d; ", msg, stack->top);
+  for (int i = 0; i < stack->top; i++) {
+    LOG("%d | ", stack->stack[i]);
+  }
+  LOG("]\n");
+}
+
+unsigned serialiseStack(ScannerStack *stack, char *buf) {
+  unsigned elements = isEmptyStack(stack) ? 0 : stack->top;
+  unsigned result_length = elements * sizeof(int);
+  int *placement = (int *)buf;
+
+  for (int i = 0; i < elements; i++) {
```

Comment:
tbf this could be `memcpy` and much faster, but I don't have intuition on how often tree-sitter serialises the stack during a normal parsing

Reply:
Sounds like a good idea.

Reply:
Addressed in 0002b5f09453f48b80405020fb76db3d6c9088e3